scipy/sparse/tests/test_base.py

```diff
@@ -248,13 +234,13 @@ class _TestCommon:
     math_dtypes = supported_dtypes
 
     # XXX: __init__ method that will be called at test collection time.
-    #      It also requires the extra warning suppression decorator.
-    #      This should be implemented in a different way.
-    @sup_sparse_efficiency2
     def __init__(self):
         # Canonical data.
         self.dat = matrix([[1,0,0,2],[3,0,1,0],[0,2,0,0]],'d')
-        self.datsp = self.spmatrix(self.dat)
+        with suppress_warnings() as sup:
+            sup.filter(SparseEfficiencyWarning,
+                       "Changing the sparsity structure of a cs[cr]_matrix is expensive")
+            self.datsp = self.spmatrix(self.dat)
```

Comment:
Rather than wrapping all the hundreds of calls to `self.spmatrix(...)` in the warning suppressor, we should just add the suppression to `self.spmatrix` in the appropriate subclasses. For example, in `TestCSR`, we could add:
```python
def spmatrix(self, *args, **kwargs):
    with suppress_warnings() as sup:
        sup.filter(...)
        return csr_matrix(*args, **kwargs)
```